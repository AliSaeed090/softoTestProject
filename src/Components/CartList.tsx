import Box from "@mui/material/Box";
import React from "react";
import {
  Grid,
   Card,
  CardContent,
  Typography,
  IconButton,
} from "@mui/material";
interface propsData {
  colour?: string;
  id?: number;
  img: string;
  name: string;
  price: number;
  Qty?: number;
  
}
interface Props {
  data: propsData;
  index:number;
  updateQty:Function
}
function CartList(props: Props) {
  let { Qty, img, name, price } = props.data;
  let index = props.index;
  const updateQty = props.updateQty;
  return (
    <Card>
      <Grid container spacing={2}>
        <Grid item xs={12} sm={1}>
          <Box
            sx={{
              display: "flex",
              justifyContent: "center",
              alignItems: "center",
            }}
          >
            <img
              style={{ width: 60, height: 100, objectFit: "cover" }}
              src={img}
              alt={"item.title"}
              loading="lazy"
            />
          </Box>
        </Grid>
        <Grid item xs={12} sm={7}>
          <Box sx={{ display: "flex", flexDirection: "column" }}>
            <CardContent sx={{ flex: "1 0 auto" }}>
              <Typography data-testid="name" component="div" variant="h6">
                {name}
              </Typography>
              <Typography
                data-testid="price"
                variant="subtitle1"
                color="text.secondary"
                component="div"
              >
                Â£{price}
              </Typography>
            </CardContent>
          </Box>
        </Grid>

        <Grid item xs={12} sm={4}>
          <CardContent sx={{ flex: "1 0 auto" }}>
            <Box
              sx={{
                display: "flex",
                justifyContent: "center",
                alignItems: "center",
              }}
            >
              <IconButton
                onClick={() => updateQty("increment", index, Qty)}
                style={{
                  padding: 20,
                  background: "#f2f2f2",
                  borderRadius: 4,
                }}
              >
                {" "}
                +{" "}
              </IconButton>
              <div style={{ display: "flex", flexDirection: "column" }}>
                <Typography
                  data-testid="Qty"
                  style={{ padding: 10, textAlign: "center" }}
                  component="div"
                  variant="subtitle1"
                >
                  {Qty}
                </Typography>
                <Typography
                  onClick={() => updateQty("remove", index, 0)}
                  style={{
                    padding: 10,
                    textAlign: "center",
                    border:"0px",
                    backgroundColor:"transparent",
                    cursor:"pointer"
                  }}
                  component="button"
                  variant="subtitle1"
                >
                  Remove
                </Typography>
              </div>

              <IconButton
                onClick={() => updateQty("decrement", index, Qty)}
                style={{
                  padding: 20,
                  background: "#f2f2f2",
                  borderRadius: 4,
                }}
              >
                {" "}
                -{" "}
              </IconButton>
            </Box>
          </CardContent>
        </Grid>
      </Grid>
    </Card>
  );
}

export default React.memo(CartList);
