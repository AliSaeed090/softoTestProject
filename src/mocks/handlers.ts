import { rest } from 'msw'
import {cartList} from './Data'
export const handlers = [
  rest.get('https://my-json-server.typicode.com/benirvingplt/products/products', (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json(cartList)
    )
  }),
]
 