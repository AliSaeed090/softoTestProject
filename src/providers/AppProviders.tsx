import { Provider } from "react-redux";
import { store } from "../redux/store";

export const AppProviders = ({ children }: { children: React.ReactNode }) => {
  return <Provider store={store}>{children}</Provider>;
};
