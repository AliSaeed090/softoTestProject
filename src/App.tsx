import React from "react";
import Pages from "../src/Pages";
import { AppProviders } from "./providers/AppProviders";

function App() {
  return (
    <AppProviders>
      <Pages />
    </AppProviders>
  );
}

export default App;
