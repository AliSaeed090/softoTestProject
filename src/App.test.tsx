import CartList from "./Components/CartList";
import App from "./Pages/index";
import { cart } from "./mocks/Data";
import { render, fireEvent, act, screen } from "./test-utils";
import renderer from "react-test-renderer";
import { AppProviders } from "./providers/AppProviders";

describe("App", () => {
  test("render App ", async () => {
    await act(async () => render(<App />));
    const textElement = screen.getByTestId("Select");
    expect(textElement).toBeInTheDocument();
  });
  test("renders a list of Products", async () => {
    await act(async () => render(<App />));
    const Products = await screen.findAllByRole("listitem");
    expect(Products).toHaveLength(3);
  });
  it("Matches App Snapshot", async () => {
    const domTree = await act(async () =>
      renderer
        .create(
          <AppProviders>
            <App />
          </AppProviders>
        )
        .toJSON()
    );

    expect(domTree).toMatchSnapshot();
  });
});
describe("CartList", () => {
  test("test Cart List Component", () => {
    const updateQty = jest.fn();
    const { getByTestId, getByRole } = render(
      <CartList data={cart} index={0} updateQty={updateQty} />
    );
    const QtyValue = Number(getByTestId("Qty").textContent);
    const nameValue = getByTestId("name").textContent;
    const priceValue = getByTestId("price").textContent;
    expect(QtyValue).toEqual(0);
    expect(nameValue).toBe("Stone Ribbed Strappy Cut Out Detail Bodycon Dress");
    expect(priceValue).toBe("Â£4");
    const incrementBtn = getByRole("button", { name: "+" });
    const decrementBtn = getByRole("button", { name: "-" });
    const removeBtn = getByRole("button", { name: "Remove" });
    fireEvent.click(incrementBtn);
    fireEvent.click(decrementBtn);
    fireEvent.click(removeBtn);
    expect(updateQty).toHaveBeenCalledTimes(3);
  });

  it("Matches Cart List Component Snapshot", () => {
    const updateQty = jest.fn();
    const domTree = renderer
      .create(<CartList data={cart} index={0} updateQty={updateQty} />)
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
