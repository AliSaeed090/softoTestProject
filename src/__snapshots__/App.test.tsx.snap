// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Matches App Snapshot 1`] = `null`;

exports[`CartList Matches Cart List Component Snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-1 css-1wztp5r-MuiGrid-root"
    >
      <div
        className="MuiBox-root css-1n2mv2k"
      >
        <img
          alt="item.title"
          loading="lazy"
          src="https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024"
          style={
            Object {
              "height": 100,
              "objectFit": "cover",
              "width": 60,
            }
          }
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-7 css-5n8ohs-MuiGrid-root"
    >
      <div
        className="MuiBox-root css-j7qwjs"
      >
        <div
          className="MuiCardContent-root css-1qaznj3-MuiCardContent-root"
        >
          <div
            className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            data-testid="name"
          >
            Stone Ribbed Strappy Cut Out Detail Bodycon Dress
          </div>
          <div
            className="MuiTypography-root MuiTypography-subtitle1 css-ecpxie-MuiTypography-root"
            data-testid="price"
          >
            Â£
            4
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-wg4xta-MuiGrid-root"
    >
      <div
        className="MuiCardContent-root css-1qaznj3-MuiCardContent-root"
      >
        <div
          className="MuiBox-root css-1n2mv2k"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "background": "#f2f2f2",
                "borderRadius": 4,
                "padding": 20,
              }
            }
            tabIndex={0}
            type="button"
          >
             
            +
             
          </button>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "column",
              }
            }
          >
            <div
              className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
              data-testid="Qty"
              style={
                Object {
                  "padding": 10,
                  "textAlign": "center",
                }
              }
            >
              0
            </div>
            <button
              className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
              onClick={[Function]}
              style={
                Object {
                  "backgroundColor": "transparent",
                  "border": "0px",
                  "cursor": "pointer",
                  "padding": 10,
                  "textAlign": "center",
                }
              }
            >
              Remove
            </button>
          </div>
          <button
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "background": "#f2f2f2",
                "borderRadius": 4,
                "padding": 20,
              }
            }
            tabIndex={0}
            type="button"
          >
             
            -
             
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
